src/str.rs

```diff
@@ -433,186 +445,229 @@ pub(crate) fn parse_str_radix_n(str: &str, radix: u32) -> Result<Decimal, crate:
+                        data += 1;
+                    }
+                    break;
+                } else {
+                    data = next;
+                    scale += passed_decimal_point as u8;
+
+                    if scale >= estimated_max_precision(RADIX) {
+                        if !bytes.is_empty() {
+                            return maybe_round_radix::<RADIX>(
+                                data as u128,
+                                bytes[0],
+                                scale,
+                                passed_decimal_point,
+                                negative,
+                            );
+                        }
+                        break;
+                    }
                 }
-                u32::from(next_byte - b'A') + 10
             }
-            b'_' => 0,
             b'.' => {
-                // Still an error if we have a second dp
-                if digits_before_dot >= 0 {
+                if passed_decimal_point {
                     return Err(Error::from("Invalid decimal: two decimal points"));
                 }
-                0
+                passed_decimal_point = true;
             }
+            b'_' => {}
             _ => return Err(Error::from("Invalid decimal: unknown character")),
-        };
-
-        // Round at midpoint
-        let midpoint = if radix & 0x1 == 1 { radix / 2 } else { (radix + 1) / 2 };
-        if digit >= midpoint {
-            let mut index = coeff.len() - 1;
-            loop {
-                let new_digit = coeff[index] + 1;
-                if new_digit <= 9 {
-                    coeff[index] = new_digit;
-                    break;
-                } else {
-                    coeff[index] = 0;
-                    if index == 0 {
-                        coeff.insert(0, 1u32);
-                        digits_before_dot += 1;
-                        coeff.pop();
-                        break;
-                    }
-                }
-                index -= 1;
-            }
         }
+        bytes = &bytes[1..];
     }
 
-    // here when no characters left
-    if coeff.is_empty() {
-        return Err(Error::from("Invalid decimal: no digits found"));
-    }
+    handle_data_radix(data, scale, negative)
+}
 
-    let mut scale = if digits_before_dot >= 0 {
-        // we had a decimal place so set the scale
-        (coeff.len() as u32) - (digits_before_dot as u32)
-    } else {
-        0
+#[inline(never)]
+#[cold]
+fn maybe_round_radix<const RADIX: u8>(
+    mut data: u128,
+    next_byte: u8,
+    scale: u8,
+    passed_decimal_point: bool,
+    negative: bool,
+) -> Result<Decimal, crate::Error> {
+    let digit = match next_byte {
+        b'0'..=b'9' => u32::from(next_byte - b'0'),
+        b'a'..=b'z' if RADIX > 10 => u32::from(next_byte - b'a' + 10),
+        b'A'..=b'Z' if RADIX > 10 => u32::from(next_byte - b'A' + 10),
+        b'_' => 0, // this is a bug?
```

Comment:
agree this looks wrong

Reply:
specifically, I guess it's not wrong, just that we shouldn't adjust scale pointlessly if we see a separator